pkg/detectors/gcp/gcp.go

```diff
@@ -90,41 +103,144 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (result
 
 		credBytes, _ := json.Marshal(creds)
 
-		s := detectors.Result{
+		s1 := detectors.Result{
 			DetectorType: detectorspb.DetectorType_GCP,
 			Raw:          raw,
 			RawV2:        credBytes,
 			Redacted:     creds.ClientEmail,
 		}
-		// Set the RotationGuideURL in the ExtraData
-		s.ExtraData = map[string]string{
+		// Set the RotationGuideURL in the ExtraData.
+		s1.ExtraData = map[string]string{
 			"rotation_guide": "https://howtorotate.com/docs/tutorials/gcp/",
 			"project":        creds.ProjectID,
 		}
 
 		if verify {
-			credentials, err := google.CredentialsFromJSON(ctx, credBytes, "https://www.googleapis.com/auth/cloud-platform")
-			if err != nil {
-				continue
-			}
-			if credentials != nil {
-				_, err = credentials.TokenSource.Token()
-				if err == nil {
-					s.Verified = true
-				}
-			}
+			client := s.getClient()
+			isVerified, verificationErr := isValidGCPServiceAccountKey(ctx, client, creds)
+			s1.Verified = isVerified
+			s1.SetVerificationError(verificationErr, creds.ClientEmail)
 		}
 
-		results = append(results, s)
+		results = append(results, s1)
 	}
 
 	return
 }
 
-func (s Scanner) IsFalsePositive(_ detectors.Result) bool {
-	return false
+// isValidGCPServiceAccountKey checks if the provided GCP service account key is valid. It verifies this by
+// comparing the public key extracted from the certificate retrieved from the `ClientX509CertURL`
+// with the public key derived from the `PrivateKey` field of the `gcpKey` struct. If the public keys
+// match, the function returns true, indicating a valid service account key. Otherwise, it returns false.
+//
+// Note: If the service account is expired or disabled, the request to fetch the certificate from the
+// `ClientX509CertURL` will fail with a 404 status code. In this case, the function will return false,
+// indicating that the service account key is invalid, regardless of the public key comparison.
+func isValidGCPServiceAccountKey(ctx context.Context, client *http.Client, key gcpKey) (bool, error) {
+	certPEM, ok, err := fetchCert(ctx, client, key.ClientX509CertURL, key.PrivateKeyID)
+	if err != nil {
+		return false, err
+	}
+	if !ok {
+		return false, nil
+	}
+
+	publicKey, ok := extractPublicKeyFromCert(certPEM)
+	if !ok {
+		return false, nil
+	}
+
+	privateKey, ok := extractPrivateKey(key.PrivateKey)
```

Comment:
Would it make sense to run this prior to the `verify` block? That way, if the key isn't valid (e.g., `---- BEGIN PRIVATE KEY ----- \n<YOUR-KEY-HERE>\n---- END PRIVATE KEY -----`) it doesn't need to make a network request. It would also filter out invalid results if verification is disabled/unverified results are being shown.